Convert RLECore from imperative to functional

saatvikshah opened the pull request:
Currently RLECore's encoding/decoding utilities are mocking imperative
for loops by using the "go" helper function which stores all necessary
state and a pseudo iteration count.

Replace the same with more functional style via composition and whole
value functions to better separate concerns and make the code more
declarative.

saatvikshah committed:
Convert RLECore from imperative to functional

Currently RLECore's encoding/decoding utilities are mocking imperative
for loops by using the "go" helper function which stores all necessary
state and a pseudo iteration count.

Replace the same with more functional style via composition and whole
value functions to better separate concerns and make the code more
declarative.

ersran9 commented on src/RLECore.hs:
```diff
@@ -1,26 +1,28 @@
-            | x == ch       = go ch (ch_count+1) result xs
-            | ch_count == 0 = go x 1 result xs
-            | otherwise     = go x 1 (result ++ [ch] ++ show ch_count) xs
+        createRunLengths :: String -> [Encoding]
```
Is it possible to rewrite this with https://hackage.haskell.org/package/groupBy-0.1.0.0/docs/Data-List-GroupBy.html#v:group ?

ersran9 replied:
Oops, wrong link - please refer to this one https://hackage.haskell.org/package/base-4.15.0.0/docs/Data-List.html#v:group

saatvikshah replied:
Thanks this is a really useful function! Done ✅

saatvikshah committed:
Use `group` to further compose RLE encoding

The initial implementation of `createRunLengthEncodings`
simultaneously(implicitly) tried to create groups of run lengths and
convert those to encodings. Obtain further separation of concerns by
splitting the same using the handy `Data.List.group`.
